openedx_learning/apps/authoring/publishing/api.py

```diff
@@ -513,3 +558,419 @@ def filter_publishable_entities(
+
+    Returns:
+        The newly created container version.
+    """
+    assert issubclass(container_version_model, ContainerVersion)
+    with atomic():
+        container = Container.objects.select_related("publishable_entity").get(pk=container_pk)
+        entity = container.publishable_entity
+        last_version = container.versioning.latest
+        assert last_version is not None
+        next_version_num = last_version.version_num + 1
+        if publishable_entities_pks is None:
+            # We're only changing metadata. Keep the same entity list.
+            next_entity_list = last_version.entity_list
+        else:
+            if entity_version_pks is None:
+                entity_version_pks = [None] * len(publishable_entities_pks)
+            next_entity_list = create_entity_list_with_rows(
+                entity_pks=publishable_entities_pks,
+                entity_version_pks=entity_version_pks,
+                learning_package_id=entity.learning_package_id,
+            )
+        next_container_version = _create_container_version(
+            container,
+            next_version_num,
+            title=title if title is not None else last_version.title,
+            entity_list=next_entity_list,
+            created=created,
+            created_by=created_by,
+            container_version_model=container_version_model,
+        )
+
+    return next_container_version
+
+
+def get_container(pk: int) -> Container:
+    """
+    [ 🛑 UNSTABLE ]
+    Get a container by its primary key.
+
+    Args:
+        pk: The primary key of the container.
+
+    Returns:
+        The container with the given primary key.
+    """
+    return Container.objects.get(pk=pk)
+
+
+@dataclass(frozen=True)
+class ContainerEntityListEntry:
+    """
+    [ 🛑 UNSTABLE ]
+    Data about a single entity in a container, e.g. a component in a unit.
+    """
+    entity_version: PublishableEntityVersion
+    pinned: bool
+
+    @property
+    def entity(self):
+        return self.entity_version.entity
+
+
+def get_entities_in_container(
+    container: Container,
+    *,
+    published: bool,
+) -> list[ContainerEntityListEntry] | None:
+    """
+    [ 🛑 UNSTABLE ]
+    Get the list of entities and their versions in the current draft or
+    published version of the given container.
+
+    Returns `None` if you request the published version and it hasn't been
+    published yet.
+
+    Args:
+        container: The Container, e.g. returned by `get_container()`
+        published: `True` if we want the published version of the container, or
+            `False` for the draft version.
+    """
+    assert isinstance(container, Container)
+    if published:
+        container_version = container.versioning.published
+        if container_version is None:
+            return None  # There is no published version of this container (yet). Should this be an exception?
```

Comment:
Yeah, I'd raise an error here for consistency for now, but if we eventually switch over to an API where they use the relation on the model, this shouldn't be something that we have to worry about.

Reply:
Sure, updated.